src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs

```diff
@@ -138,7 +133,13 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                 // used to name them.  The boilerplate code generation that happens here is minimal when compared to
                 // the work required to generate the actual matching code for the regex.
                 int id = 0;
-                string generatedClassName = $"__{ComputeStringHash(compilationDataAndResults.Right.AssemblyName ?? ""):x}";
+
+                // We put all of the file types into a custom namespace, just to minimize the chances of there being a
+                // conflict between the generated type names and something defined in the user's code.  Having the generated
+                // types be "file" means the type definitions themselves won't conflict nor be usable by user code (yay),
+                // but if user code used the same the full type names, the call sites we have to access the file types might
+                // become ambiguous and end up preferring the user-defined types (boo).
```

Comment:
@jaredpar, @RikkiGibson, is there a better solution here? i.e. is there any way to enable code defined in a partial class in the generated code to bind more tightly to a file-scoped class in the same file than to a class the user code may have defined?

Reply:
The behavior you're seeing is a bug. We do expect the file type to fully "shadow" the non-file type here. I'll take a look.

Reply:
Got it now. The key here is that where `Print()` is implemented, `UserCode.C` is in a nearer scope than `C@OtherFile`. If you access `C` as `global::C` then you get the behavior you want.

It wasn't clear to me if introducing a `file` type should only shadow types with the same name and namespace, or if it should shadow any type with the same name. I'll dig into that and follow up.

Reply:
> If you access C as global::C then you get the behavior you want.

Ah, ok.  So I can just get rid of the namespace around the file-scoped types entirely, use global::ClassName to refer to them in the same file, and that'll do the right thing?  If so, great.